File: src/core/exceptions.ts
/**
 * Custom exceptions for ETL pipeline operations.
 */

export class ExtractionFailedException extends Error {
  constructor(message?: string) {
    super(message ? `Extraction failed: ${message}` : "Extraction failed");
    this.name = "ExtractionFailedException";
  }
}

export class TransformFailedException extends Error {
  constructor(message?: string) {
    super(message ? `Transform failed: ${message}` : "Transform failed");
    this.name = "TransformFailedException";
  }
}

export class UploadFailedException extends Error {
  constructor(message?: string) {
    super(message ? `Upload failed: ${message}` : "Upload failed");
    this.name = "UploadFailedException";
  }
}

export class UnknownDataPatternException extends Error {
  unknownValue: string;
  builderClass: string;

  constructor(unknownValue: string, builderClass: string, message?: string) {
    super(
      message ??
        `Unknown data pattern in ${builderClass}: ${unknownValue}`,
    );
    this.name = "UnknownDataPatternException";
    this.unknownValue = unknownValue;
    this.builderClass = builderClass;
  }
}

export class ArchiveProcessingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ArchiveProcessingError";
  }
}

export class UnsupportedProviderError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnsupportedProviderError";
  }
}

File: context_use/core/exceptions.py
"""Custom exceptions for ETL pipeline operations."""


class ExtractionFailedException(Exception):
    def __init__(self, message: str | None = None):
        self.message = (
            f"Extraction failed: {message}" if message else "Extraction failed"
        )
        super().__init__(self.message)


class TransformFailedException(Exception):
    def __init__(self, message: str | None = None):
        self.message = f"Transform failed: {message}" if message else "Transform failed"
        super().__init__(self.message)


class UploadFailedException(Exception):
    def __init__(self, message: str | None = None):
        self.message = f"Upload failed: {message}" if message else "Upload failed"
        super().__init__(self.message)


class UnknownDataPatternException(Exception):
    """Raised when a ThreadPayloadBuilder encounters unknown/unmatched data pattern."""

    def __init__(
        self,
        unknown_value: str,
        builder_class: str,
        message: str | None = None,
    ):
        self.unknown_value = unknown_value
        self.builder_class = builder_class
        self.message = (
            message or f"Unknown data pattern in {builder_class}: {unknown_value}"
        )
        super().__init__(self.message)


class ArchiveProcessingError(Exception):
    """Top-level error for the process_archive entry point."""

    pass


class UnsupportedProviderError(ValueError):
    """Raised when an unknown provider is requested."""

    pass
